tools/run_tests/xds_k8s_test_driver/framework/infrastructure/gcp/api.py

```diff
@@ -260,29 +261,68 @@ class OperationError(Error):
+    error: status_pb2.Status
 
-    def __init__(self, api_name, operation_response, message=None):
+    def __init__(self, api_name: str, response: dict):
         self.api_name = api_name
-        operation = json_format.ParseDict(
-            operation_response,
-            Operation(),
-            ignore_unknown_fields=True,
-            descriptor_pool=error_details_pb2.DESCRIPTOR.pool)
+
+        # Operation.metadata field is Any specific to the API. It may not be
+        # present in the default descriptor pool, and that's expected.
+        # To avoid json_format.ParseError, handle it separately.
+        self.metadata = response.pop('metadata', {})
+
+        # Must be after removing metadata field.
+        operation: Operation = self._parse_operation_response(response)
         self.name = operation.name or 'unknown'
         self.code_name = code_pb2.Code.Name(operation.error.code)
         self.error = operation.error
-        # Collect error details packed as Any without parsing concrete types.
-        self.error_details = [
-            text_format.MessageToString(any_error, as_one_line=True)
-            for any_error in self.error.details
-        ]
-        if message is None:
-            message = (f'{api_name} operation "{self.name}" failed. Error '
-                       f'code: {self.error.code} ({self.code_name}), '
-                       f'message: {self.error.message}, '
-                       f'details: {self.error_details}')
-        self.message = message
-        super().__init__(message)
+        super().__init__()
+
+    @staticmethod
+    def _parse_operation_response(operation_response: dict) -> Operation:
+        try:
+            return json_format.ParseDict(
+                operation_response,
+                Operation(),
+                ignore_unknown_fields=True,
+                descriptor_pool=error_details_pb2.DESCRIPTOR.pool)
+        except (json_format.Error, TypeError) as e:
+            # Swallow parsing errors if any. Building correct OperationError()
+            # is more important than losing debug information. Details still
+            # can be extracted from the warning.
+            logger.warning(
+                ("Can't parse response while processing OperationError: '%r', "
+                 "error %r"), operation_response, e)
+            return Operation()
+
+    def __str__(self):
+        indent_l1 = ' ' * 2
+        indent_l2 = indent_l1 * 2
+
+        result = (f'{self.api_name} operation "{self.name}" failed.\n'
+                  f'{indent_l1}code: {self.error.code} ({self.code_name})\n'
+                  f'{indent_l1}message: "{self.error.message}"')
+
+        if self.error.details:
+            result += f'\n{indent_l1}details: [\n'
+            for any_error in self.error.details:
+                error_str = json_format.MessageToJson(any_error)
+                for line in error_str.splitlines():
+                    result += indent_l2 + line + '\n'
```

Comment:
Do you think we need double line breaks here? Otherwise, we might not able to separate one error from another, since text proto format is a bit weird (no outer parenthesis).

Reply:
By default, `splitlines()` removes line breaks, I'm re-adding them back.

Reply:
Good point, I want to comment on the space between error messages. See comment on :316.